enhancement(file sink): Add support for gzip compression

perfectayush opened the pull request:
File sink currently doesn't support compression, this rudimentary change will
allow to use compression in file sink

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

Commit:
enhancement(file sink): Add support for gzip compression

File sink currently doesn't support compression, this rudimentary change will
allow to use compression in file sink

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

perfectayush commented:
This is a WIP for #3365. Currently, doesn't have any associated tests.

ktff reviewed (commented):
@perfectayush I left a few comments, but it's on the right track. I'm curious, what was failing with `Box<dyn AsyncWriteExt>`? If we could somehow use it, we could remove half of the code.

perfectayush commented on src/sinks/file/mod.rs:
```diff
@@ -86,6 +127,10 @@ impl FileSink {
             encoding: config.encoding.clone(),
             idle_timeout: Duration::from_secs(config.idle_timeout_secs.unwrap_or(30)),
             files: ExpiringHashMap::default(),
+            compression: match config.compression {
+                Compression::Gzip => true,
+                Compression::None => false,
+            },
```
Sure, let me try by using compression directly instead of making this a bool.

perfectayush commented on src/sinks/file/mod.rs:
```diff
@@ -134,7 +179,7 @@ impl FileSink {
                             // We got an expired file. All we really want is to
                             // flush and close it.
-                            if let Err(error) = expired_file.flush().await {
+                            if let Err(error) = expired_file.shutdown().await {
```
Yeah, I had written some test code with async_compression library and just using `flush`, the gzip file was not being written properly. Using shutdown, however, was closing the file properly. I had guessed that using shutdown for GzipEncoder flushes with required gzip headers, which are not proper with just `flush`.
I havn't dug in async_compression code so not sure.

bruceg replied:
FTR a `flush` should just ensure whatever is buffered is written out to the compressed stream, but that won't necessarily ensure that the file is closed off properly, which involves fixing up the header and checksums. The latter should only ever be done once, hence a separate `shutdown`.

perfectayush replied:
I just realize that `shutdown` requires to be called on all files, especially in case of `GzipEncoder` when vector shuts down. Else the required gzip headers won't be written. 

I am not sure how/where this shutdown handling will be added.

ktff replied:
Also, I went through `tokio` code and, for `File` `shutdown` won't flush the buffers, it just returns `Ok`. This is a bug since it violates `AsyncWrite` contract. https://github.com/tokio-rs/tokio/issues/2760 

We should call `flush` and `shutdown` until that issue is resolved, and link the issue in the comment.

ktff replied:
Nope disregard my comment before, all is fine.

perfectayush commented on src/sinks/file/mod.rs:
```diff
@@ -53,6 +60,39 @@ impl Default for Encoding {
     }
 }
 
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone)]
```
Ack.

perfectayush commented:
@ktff The problem I was running into while using Box\<dyn AsyncTrait\> was, I was getting error along the lines of:
```
the trait cannot be made into an object because method `write` references the `Self` type in its return type
```
This was with respect to all methods in trait `AsyncWriteExt`'s  referencing Self in their definition. There were quite a few of those.  I wasn't sure on how to resolve that and hit a dead end.

Commit:
Fix type for compression field in FileSink struct

FileSink struct's compression field was earlier initialized as a bool. Using
Compression enum directly is a better abstraction, simplifies extending FileSink
for other compression types in future

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

ktff commented:
A, it's not [object safe](https://doc.rust-lang.org/book/ch17-02-trait-objects.html#object-safety-is-required-for-trait-objects), shame.

binarylogic commented:
I'm curious how this works in practice? Vector will write multiple times to a single file which means gzip must support streaming compression. And if it does, I assume compression is reduced since it is compressing chunks instead of the whole file. This is partly why we recommend `logrotate` since it'll rotate and compression the entire file in one shot.

perfectayush commented:
@binarylogic as per [this](https://stackoverflow.com/questions/19700503/how-does-gzip-compression-rate-change-when-streaming-data) stackoverflow entry, 

> gzip algorithm uses a sliding window of the last 32K of data in which it searches for matching strings.

I generated a few gzip file using the generator source toml configuration shared above, it resulted in one gzipped file of size ~ 9 MiB on average per minute. Uncompressing using `gunzip` yielded a file of ~ 180 MiB. Gzipping back using `gzip` yielded a file of ~ 9 MiB again, even with best compression level of `-9`.  (Btw tested on macOS, but should be same on linux as well I think, unless there is much difference in gzip implementation)

Since logrotate also uses `gzip` utility, the compression will be of similar level. I have concerns wrt gzipping logs via logrotate. Implementing a logrotate config which does not affect a file currently being written to by vector will be hard. Since vector will write all files in a single directory and rotation is already being taken care of by vector templating, there is little need of logrotate here. 

Also doing streaming gzip will result in reduced cpu utilization spread over a period of time, instead of spiky cpu utilization via logrotate in batch gzip mode.

As far as a real life use-case is concerned, we have in past sent logs to a centralized log server over td-agent protocol and stored gzipped files on EBS volume. This is useful for using `zgrep` to grep the files. Gzipping ensures reduced disk utilization. Writing file with timestamp templates helps us cleaning up old log file using `find -mtime -delete`.

perfectayush commented:
@ktff Thanks for pointing to the object safety docs. I didn't know that. I think using a macro is another way to reduce code. But I am not familiar with the rust's macro system and don't want to mess with it, right now.

binarylogic commented:
Sounds goood to me. I was not aware that Gzip worked on 32k at a time. If the resulting size is similar then this is indeed better than logrotate.

bruceg commented:
> Sounds goood to me. I was not aware that Gzip worked on 32k at a time. If the resulting size is similar then this is indeed better than logrotate.

I don't think this is accurate. Gzip does not work on a block at a time, it is streaming, and I don't think that is what is being claimed here. By default, gzip will buffer internally such that the multiple small writes have the same effect as the expected streaming. Only by explicitly flushing compression will gzip/zlib explicitly compress and write out that buffer early. This means there should be no size benefit to re-compressing after the fact (ie with logrotate) over doing it internally. It does come at a cost, though, as it also means that writes to the files may be delayed until those internal buffers (both pre- and post-compression) happen to be flushed or when vector is done (at shutdown or close) increasing the risk of lost data.

bruceg commented on src/sinks/file/mod.rs:
```diff
@@ -53,6 +60,46 @@ impl Default for Encoding {
     }
 }
 
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy)]
+#[serde(rename_all = "snake_case")]
+pub enum Compression {
+    Gzip,
+    None,
+}
```
Note that there already exists an `enum Compression` in `src/sinks/util/buffer/mod.rs`, which matches this definition almost exactly. Would that be usable here to avoid the duplicate definition?

perfectayush replied:
I had noticed that. It is almost similar, but I felt that implementation more tied to flate2/std::io::Write. 
While this one relies on async_compression and async writes via tokio. With async_compression, it can be quickly extended for other compression types if required.

perfectayush commented:
So there is a bug in this code, when vector shutsdown (via a SIGTERM or SIGINT), all the files currently being written to need to call AsyncWriteExt's `shutdown` method to write Gzip headers. 

This is not an issue with Regular files, as they do not require flushing of headers.

One way to do this is to expire all the keys in ExpiringHashMap and iterate over them and `shutdown` all the files. I don't think ExpiringHashMap currently exposes an API to expire all the keys, or even iterate over them. I have identified that this needs to happen in the `None` branch for input event. Please suggest the right approach.

https://github.com/timberio/vector/blob/6bac9f1c6bc74f2b775f5a18ea25b3796f58a555/src/sinks/file/mod.rs#L118-L128

ktff commented:
> all the files currently being written to need to call AsyncWriteExt's shutdown method to write Gzip headers.

Good catch.

> One way to do this is to expire all the keys in ExpiringHashMap and iterate over them and shutdown all the files. 

Yep, this is the way.

> I don't think ExpiringHashMap currently exposes an API to expire all the keys, or even iterate over them.

True it doesn't, but `ExpiringHashMap` is in our code base so we can extend it with a method like
```
pub fn take(self)->HashMap<_,_>{
...
}
```
With that you have all the pieces, just make sure to call both `flush` and `shutdown` as mentioned in https://github.com/timberio/vector/pull/3373#discussion_r468527103

ktff commented:
Going with what @perfectayush said, I think greatest value of this feature is reduction of io-disk utilization which is substantial considering not only that we are writing once instead of reading once and writing twice, but also that that the write is also smaller. Going with 9MB to 180MB example, it would be a `97.5%` reduction in io. So it has value even if it's less efficiente than `logrotate`.

Commit:
Flush files at shutdown in file sink

File sink should flush files at shutdown not just when files are expired. Not
doing so may result in data loss.

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

Commit:
Make linters and formatters happy

* Move `use async_compression` to recommended place by rustfmt
* Fix whitespace

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

perfectayush commented:
I have added the required lint changes and added shutdown handling. I have tested shutdown handling both on linux and mac.

Sometimes, when starting vector with generator code above, I am getting error:
```
Aug 12 13:58:10.665 ERROR vector::sinks::file: Failed to flush file. path=b"logtest-2020-08-12-58.log.gz" error=task was cancelled
```

I notice in these situations, the log line is not proper in the gzip file. Can you suggest what needs to be fixed. My guess is tokio thread is closing before flush happens properly. But I am not familiar enough with codebase to understand what needs to be fixed.

ktff commented:
That's a bug in `vector` issue #3438. 

This feature requires fix for #3438 to be usable, so it's blocked on that.

ktff commented:
@perfectayush it's fixed, you can merge the master to this branch and double check that it works. 

Regarding tests, one is sufficient. It would go something like this:
* Construct `file` sink
* Run `sink` with random data, `test_util` module has functions for generating it
* Load the file and run it through GzipDecoder
* Assert that the random data we have given to the sink is the same we got from file 
* Rejoice

Commit:
Add tests for gzip file sink

Signed-off-by: Ayush Goyal <ayush@helpshift.com>

Commit:
Merge branch 'master' into support-gzip-file-sink

* master: (47 commits)
  chore(topology): Wait for detached sink to finish  (#3450)
  fix(sinks): Wrap StreamSink around TcpSink (#3495)
  chore: remove tower:0.1 (#3481)
  docs(http sink): Fix default for rate_limit_num (#3498)
  chore: update tokio in tls (#3449)
  fix(prometheus source): rewrite parser, improve error handlings  (#3286)
  chore(fix): OSX doesn't have realpath - added it via coreutils to the bootstrap (#3489)
  chore: Minor spelling, grammar and style fixes (#3472)
  chore: update aws_cloudwatch_logs (#3422)
  chore(ci): do not cancel builds on master (#3462)
  chore(ci): Update the target Ubuntu distros for packaging (#3484)
  enhancement(ci): Ship Helm Chart from release CI workflow (#3459)
  chore: remove bytes:0.4 (#3460)
  chore: Fix stoped typo (#3470)
  chore: Update typo in Docker internal event logging
  chore: Remove ARMv7 references
  enhancement(file source): Even more instrumentation (#3436)
  chore: Extract line_agg from file source (#3451)
  fix: doc test in event/metric (#3461)
  enhancement(console sink): Add "text" encoding for metrics (#3434)
  ...

perfectayush commented:
Thanks @ktff. I have added a simple test and it's working.

I have tested after merging master and it's working as intended now.

ktff reviewed (approved):
Nice work. 

Regarding the test failing on windows, it's the same race in the test as #2155. It will be fixed with #3475 which has an overlap with this PR so I would like to merge this one first.

@binarylogic docs need to be added.

ktff commented:
I'll merge this now, but the docs need to be added. 

@perfectayush if you are up for it, this `compression` needs to be documented here https://github.com/timberio/vector/blob/master/.meta/sinks/file.toml.erb. A template version is here https://github.com/timberio/vector/blob/master/.meta/_partials/fields/_compression_options.toml.erb that the other sinks use, it's mostly usable only `description` is not correct. It just looks complicated, at least my first time was.
